server/src/main/java/org/elasticsearch/rest/RestController.java

```diff
@@ -104,25 +107,34 @@ public RestController(Set<RestHeaderDefinition> headersToCopy, UnaryOperator<Res
      * @param handler The handler to actually execute
      * @param deprecationMessage The message to log and send as a header in the response
      */
-    protected void registerAsDeprecatedHandler(RestRequest.Method method, String path, RestHandler handler, String deprecationMessage) {
+    protected void registerAsDeprecatedHandler(Method method, String path, RestApiVersion version,
+                                               RestHandler handler, String deprecationMessage) {
         assert (handler instanceof DeprecationRestHandler) == false;
-
-        registerHandler(method, path, new DeprecationRestHandler(handler, deprecationMessage, deprecationLogger));
+        if (version.major == RestApiVersion.current().major) {
```

Comment:
would it be better we actually use `version == RestApiVersion.current()` instead of `.major` ? Those are enums, so comparison should look good I think.
THe same for `version.major < RestApiVersion.current().major` Are we expecting a version to be ever different then minimumSupported here (N-1) ? how about we just use `version == RestApiVersion.minimuSupported`

With that we would prevent registering N-2 handlers after the version bump was performed, but N-2 compatible handlers were not removed yet.